styx-common/src/main/java/com/spotify/styx/api/Middlewares.java

```diff
@@ -126,26 +133,62 @@ private Middlewares() {
     };
   }
 
-  public static <T> Middleware<AsyncHandler<Response<T>>, AsyncHandler<Response<T>>> exceptionHandler() {
+  public static <T> Middleware<AsyncHandler<Response<T>>, AsyncHandler<Response<T>>> exceptionAndRequestIdHandler() {
     return innerHandler -> requestContext -> {
+
+      // Accept the request id from the incoming request if present. Otherwise generate one.
+      final String requestIdHeader = requestContext.request().headers().get(X_REQUEST_ID);
+      final String requestId = (requestIdHeader != null)
+          ? requestIdHeader
+          : UUID.randomUUID().toString().replace("-",""); // UUID with no dashes, easier to deal with
+
+      MDC.put(REQUEST_ID, requestId);
+
       try {
         return innerHandler.invoke(requestContext).handle((r, t) -> {
+          MDC.remove(REQUEST_ID);
```

Comment:
Isn't it easier and more robust to remove it once, in `finally`? There's even a try-with-resource example in https://gquintana.github.io/2017/12/01/Structured-logging-with-SL-FJ-and-Logback.html.

Reply:
This removal inside the async handler was misplaced. Good catch. 

Didn't know about the try-with-resource helper. Will try to use that instead.